libpod/image/layer_tree.go

```diff
@@ -0,0 +1,211 @@
+package image
+
+import (
+	"context"
+
+	ociv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+// layerTree is an internal representation of local layers.
+type layerTree struct {
+	// nodes is the actual layer tree.
+	nodes map[string]*layerNode
+	// ociCache is cache for Image.ID -> OCI Image.
+	ociCache map[string]*ociv1.Image
+}
+
+// node returns a layerNode for the specified layer.
+func (t *layerTree) node(layer string) *layerNode {
+	node, exists := t.nodes[layer]
+	if !exists {
+		node = &layerNode{}
+		t.nodes[layer] = node
+	}
+	return node
+}
+
+// toOCI returns an OCI image for the specified image.
+func (t *layerTree) toOCI(ctx context.Context, i *Image) (*ociv1.Image, error) {
+	var err error
+	oci, exists := t.ociCache[i.ID()]
+	if !exists {
+		oci, err = i.ociv1Image(ctx)
+		t.ociCache[i.ID()] = oci
+	}
+	return oci, err
+}
+
+// layerNode is a node in a layerTree.  It's ID is the key in a layerTree.
+type layerNode struct {
+	children []*layerNode
+	images   []*Image
+	parent   *layerNode
+}
+
+// layerTree extracts a layerTree from the layers in the local storage and
+// relates them to the specified images.
+func (ir *Runtime) layerTree() (*layerTree, error) {
```

Comment:
It would be nice if `GetLayersMapWithImageInfo` were rebuilt on top of this. OTOH, sure, not in scope.